### ISSUE TITLE
Question: Understanding predictions and distances

### ISSUE THREAD

**Issue by Mohinta2892** [AAitorG/napari-dinoSim#15]
Hi @AAitorG,

Sorry to bother you so much, but I am trying to leverage dinosim for some datasets on our end and have a few general questions.

I am currently only trying to build the `ref` embeddings for a disparate set of images (2D).
I have created the below function which I can call within a loop:

```bash
def call_dinosim(few_shot, train_dataset, points, crop_shape=(512, 512, 1), upsample="bilinear"):
    # upsample = "bilinear" #@param {type:"string", options:["bilinear", "Nearest Neighbor", "None"], value-map:{bilinear:"bilinear", "Nearest Neighbor": "nearest", None:None}}
    # crop_shape = (512,512,1) #@param {type:"raw"}

    kernel = gaussian_kernel(size=3, sigma=1)
    kernel = torch.tensor(kernel, dtype=torch.float32, device=device)
    filter_f = lambda x: torch_convolve(x, kernel)

    few_shot.delete_precomputed_embeddings() # must do this for every independent reference
    few_shot.delete_references()
    if not few_shot.emb_precomputed:
        few_shot.pre_compute_embeddings(train_dataset, overlap=(0, 0), padding=(0, 0), crop_shape=crop_shape,
                                        verbose=True, batch_size=5)
    few_shot.set_reference_vector(list_coords=points)
    distances = few_shot.get_ds_distances_sameRef(verbose=True)
    predictions = few_shot.distance_post_processing(distances, filter_f, upsampling_mode=upsample)
    print("Finished")

    return predictions, distances
```

I have significantly resized (Cropped) my input images which are now `1,40,40,1` in shape. I have changed the `crop_shape=(64,64,1)` and the `resize_size=56` (a multiple of patch_size=14).

So the predictions returned are in the `original input image shape = 1,40,40,1`
and the distances are calculated between `patches` of the original image, which ends up to be `1,4,4` (56/14 = 4, per my understanding).

Few questions now:

- When I match my inference to its respective ref (every image is different and not of the same object), I should remove the below two lines, right? This should use the existing embedding of the last image and check with an inference?

```
    few_shot.delete_precomputed_embeddings() # must do this for every independent reference
    few_shot.delete_references()
```

Do we still have to run this at inference time?
I assume no as per the paper? Is there an example of a creating a reference (with load/save capabilities in the pipeline) and then use it as the same reference for an inference image?

```
few_shot.set_reference_vector(list_coords=points)
```

-  How best to understand the distances? Are these patches always sequential? For example, given a 56,56 image, the patch 0 correspond to 0:14, 0:14 in the resized image space? 

- Can we plot the distances as a heatmap and that can tell us the semantic similarity in the patches of the whole image ? Something like below (forgive my plotting skills). How will this translate when the similarities are between `ref` and `inference` images?

![Image](https://github.com/user-attachments/assets/17c211b7-0fd0-4619-9604-2ca75678fc3f)

- What is the implication of ` batch_size=5`? I assume this just batches through the train_dataset to create the embeddings and is purely for computational efficiency. So changing it to 1 should not affect me at all, given currently I have 1 image per independent reference category?

Thanks v much in advance and please forgive my naive questions.

Best,
Samia

**Comment by AAitorG** [comment]
Hi @Mohinta2892 ,

I'll start with some general notes about the pipeline, and then I'll go through your questions.

## Understanding the Pipeline Architecture

### Image Processing Workflow

The `crop_shape` parameter defines how your input image gets divided into smaller sections for processing. Without padding and overlap (as shown in your implementation), the operation looks something like this:

![Image](https://github.com/user-attachments/assets/eacafa37-f61f-4dec-ba90-bd88abc897ba)

Here's how the processing chain works:

**Processing Steps:**
1. **Initial Cropping**: Your image gets divided into crops based on `crop_shape`
2. **Resizing**: Each crop is resized to `resize_size` before being fed to DINOv2
3. **Tokenization**: DINOv2 processes each crop individually, internally dividing it into 14×14 pixel patches (tokens)
4. **Reconstruction**: Results are recombined and scaled back to original image dimensions

**Example with your parameters:**
- Input image: `(1, 256, 256, 1)` *(batch, height, width, channels)*
- Generated crops: `(16, 64, 64, 1)` 
- Resized for DINOv2: `(16, 56, 56, 1)`
- Token representation: `(16, 4, 4, 1)` *(since 56/14 = 4)*
- Final output: Reconstructed to original image shape

### Scale Considerations

The effective scale factor in your setup is 56/64 = 0.875, meaning DINOv2 sees a slightly compressed version of each crop. For detecting smaller objects or fine details, consider increasing the `resize_size`. The model performs well with images around 518×518 pixels. I suggest you to experiment with `crop_shape` to optimize for your specific object scales.

### Multi-Image Processing

The system supports processing multiple images simultaneously within the same dataset. The first dimension specifies the image count, eliminating the need for manual iteration. When setting reference coordinates, the z-coordinate indicates which image the point belongs to:

```python
# Example: First point from image 0, second point from image 2
points = [(0, 420, 180), (2, 123, 321)]  # (z, x, y) format
```

Once reference is set, you can run:
```python
distances = few_shot.get_ds_distances_sameRef()
predictions = few_shot.distance_post_processing(distances, ...)
```
These functions compute distances and generate predictions for all precomputed images in the dataset using the same reference. This allows you to detect the presence of the object of interest (as defined by the reference points) across all precomputed images.

## Addressing Your Specific Questions

### 1. Embedding Management for Independent References

You're correct about removing those deletion lines when working with the same embeddings. The `pre_compute_embeddings()` function only needs to run once if your parameters remain unchanged.

I recommend you to delete embeddings only when switching to a different set of images.

### 2. Reference Persistence and Inference

The system provides built-in save/load functionality for both references and embeddings:

**For references:**
```python
# Save current reference
few_shot.save_reference("./references/obj_reference.pt")

# Load existing reference
few_shot.load_reference("./references/obj_reference.pt")
```

**For embeddings:**
```python
# Save computed embeddings
few_shot.save_embeddings("./embeddings/img_embedding.pt")

# Load existing embeddings
few_shot.load_embeddings("./embeddings/img_embedding.pt")
```

The reference file contains complete reference information and can work with different embedding sets, allowing you to use references across different image collections.

### 3. Understanding Distance Maps

Distance values correspond to sequential image patches following Vision Transformer (ViT) architecture patterns. For a 56×56 input, patches map as:
- Patch 0: `[0:14, 0:14]` *(x, y)*
- Patch 1: `[14:28, 0:14]`
- And so forth...

These distances represent semantic similarity between image patches and your reference. Visualizing them as heatmaps reveals regions with similar semantic content. You can visualize it at full size by plotting `predictions`.

### 4. Batch Size Impact

Your understanding is correct. The `batch_size` parameter only affects computational efficiency by controlling how many crops are processed simultaneously by DINOv2. Setting it to 1 won't change your results.

Let me know if anything’s unclear or if you have further questions. I'm happy to help!

Best,
Aitor
